include/networkit/graph/Graph.hpp

```diff
@@ -139,6 +145,379 @@ class Graph final {
     //!< same schema (and same order!) as outEdges
     std::vector<std::vector<edgeid>> outEdgeIds;
 
+private:
+    // base class for all node attribute storages with attribute type info
+    // independent of the attribute type, holds bookkeeping info only:
+    // - attribute name
+    // - type info of derived (real storage holding) classes
+    // - which indices are valid
+    // - number of valid indices
+    // - the associated graph (who knows, which nodes exist)
+    // - the validity of the whole storage (initially true, false after detach)
+    class NodeAttributeStorageBase {
+    public:
+        NodeAttributeStorageBase(const Graph *graph, std::string name, std::type_index type)
+            : name{std::move(name)}, type{type}, theGraph{graph}, validStorage{true} {}
+
+        void invalidateStorage() { validStorage = false; }
+
+        const std::string &getName() const noexcept { return name; }
+
+        std::type_index getType() const noexcept { return type; }
+
+        bool isValid(node n) const noexcept { return n < valid.size() && valid[n]; }
+
+        // Called by Graph when node n is deleted.
+        void invalidate(node n) {
+            if (isValid(n)) {
+                valid[n] = false;
+                --validElements;
+            }
+        }
+
+    protected:
+        void markValid(node n) {
+            if (!theGraph->hasNode(n))
+                throw std::runtime_error("This node does not exist");
+
+            if (n >= valid.size())
+                valid.resize(n + 1);
+
+            if (!valid[n]) {
+                valid[n] = true;
+                ++validElements;
+            }
+        }
+
+        void checkIndex(node n) const {
+            if (!theGraph->hasNode(n)) {
+                throw std::runtime_error("This node does not exist");
+            }
+            if (!isValid(n)) {
+                throw std::runtime_error("Invalid attribute value");
+            }
+        }
+
+    private:
+        std::string name;
+        std::type_index type;
+        std::vector<bool> valid; // For each node: whether attribute is set or not.
+    protected:
+        index validElements = 0;
+        const Graph *theGraph;
+        bool validStorage; // Validity of the whole storage
+
+    }; // class NodeAttributeStorageBase
+
+    template <typename T>
+    class NodeAttribute;
+
+    template <typename T>
+    class NodeAttributeStorage : public NodeAttributeStorageBase {
+    public:
+        NodeAttributeStorage(const Graph *theGraph, std::string name)
+            : NodeAttributeStorageBase{theGraph, std::move(name), typeid(T)} {}
+
+        void resize(node i) {
```

Comment:
Is there a reason why `resize` and `size` are implemented in the derived class and not in the base class?

Reply:
Because the attribute values are stored in the derived class (`values` vector).

Reply:
Ah, right - somehow overlooked this. And the whole approach of having a base class is because of the generalized `NodeAttributeMap`. Ok, nevermind then.